python/lsst/ip/isr/linearize.py

```diff
@@ -26,226 +26,510 @@
+    corrected_value = uncorrected_value + f(uncorrected_value)
+    """
+    LinearityType = None  # linearity type, a string used for AmpInfoCatalogs
 
-        @throw RuntimeError if anything doesn't match
+    @abc.abstractmethod
+    def __call__(self, image, **kwargs):
+        """Correct non-linearity.
+
+        Parameters
+        ----------
+        image : `lsst.afw.image.Image`
+            Image to be corrected
+        kwargs : `dict`
+            Dictionary of parameter keywords:
+            ``"coeffs"``
+                Coefficient vector (`list` or `numpy.array`).
+            ``"table"``
+                Lookup table data (`numpy.array`).
+            ``"log"``
+                Logger to handle messages (`lsst.log.Log`).
+
+        Returns
+        -------
+        output : `bool`
+            If true, a correction was applied successfully.
+
+        Raises
+        ------
+        RuntimeError:
+            Raised if the linearity type listed in the
+            detector does not match the class type.
         """
-        ampInfoType = detector.getAmplifiers()[0].getLinearityType()
-        if self.LinearityType != ampInfoType:
-            raise RuntimeError("Linearity types don't match: %s != %s" % (self.LinearityType, ampInfoType))
+        pass
 
 
 class LinearizeLookupTable(LinearizeBase):
-    """Correct non-linearity with a persisted lookup table
+    """Correct non-linearity with a persisted lookup table.
 
     for each i,j of image:
         rowInd = int(c0)
         colInd = int(c1 + uncorrImage[i,j])
         corrImage[i,j] = uncorrImage[i,j] + table[rowInd, colInd]
 
-    where c0, c1 are collimation coefficients from the AmpInfoTable of the detector:
-    - c0: row index; used to identify which row of the table to use (typically one per amplifier,
-            though one can have multiple amplifiers use the same table)
-    - c1: column index offset; added to the uncorrected image value before truncation;
-            this supports tables that can handle negative image values; also, if the c1 ends with .5
-            then the nearest index is used instead of truncating to the next smaller index
+    where c0, c1 are collimation coefficients from the detector amplifier:
+    - c0: row index; used to identify which row of the table to use
+            (typically one per amplifier, though one can have multiple
+            amplifiers use the same table)
+    - c1: column index offset; added to the uncorrected image value
+            before truncation; this supports tables that can handle
+            negative image values; also, if the c1 ends with .5 then
+            the nearest index is used instead of truncating to the
+            next smaller index
 
-    In order to keep related data together, the coefficients are persisted along with the table.
-    """
-    LinearityType = "LookupTable"
+    In order to keep related data together, the coefficients are
+    persisted along with the table.
```

Comment:
This docstring seems to come from the old code, correct? Does it still describe properly what your new code does?  I have some questions (some of them actually come from t things I quite did not understand the first time I read the docstring in the old code): 

-  "for each i,j of image"? which image? The amplifier image? (I believe so, but I think it's better to be explicit in the docstring). 
-  "c0, c1 are collimation coefficients": what does "collimation" exactly mean in this context? 
-  are c0 and c1, in general, not necessarily integers?
- "c1" is a columns index or the correction from the LUT?
- "In order to keep related data together, the coefficients are persisted along with the table.": is this still true?

Reply:
I've updated the docstring, and moved relevant portions to the Linearizer init docstring.  For your questions:
* The LinearizeBase (of which LinearizeLookupTable is a subclass) operate on a complete afw Image.  The Linearizer code handles iteration over amplifier BBoxes, and passes a view onto the amplifiers' bbox to the LinearizeBase math code.  This allows each of these functions to simply do the math, and not need to worry about amplifiers/bboxes/etc.
* "collimation" is used here as "selecting a column from the table."  c0 is used to select a row in the lookup table for this amplifier, and c1 as an offset in the lookup table vector pointed to by c0.
* c0 and c1 are integers, and the image value is also an integer for lookup tables.  c0 gives the row to use, and (imVal + c1) gives the column.  That entry in the table yields the amount to add to the image.
* Line removed as I don't know that it was ever true.